aegisthus-hadoop/src/main/java/org/coursera/mapreducer/CQLMapper.java

```diff
@@ -148,42 +144,101 @@ private void handleGroup(Context context, ColumnGroupMap group, ByteBuffer[] key
-            throw new RuntimeException("Collections not supported yet.");
+            List<Pair<ByteBuffer, Column>> collection = group.getCollection(name.name.key);
+            ByteBuffer buffer = ((CollectionType)name.type).serialize(collection);
+            addCqlCollectionToRecord(record, name, buffer);
         } else {
             Column c = group.getSimple(name.name.key);
             addCqlValueToRecord(record, name, (c == null) ? null : c.value());
         }
     }
 
-    private void addCqlValueToRecord(GenericRecord record, CFDefinition.Name name, ByteBuffer value) {
-        if (value == null) {
-            record.put(name.name.toString(), null);
-            return;
+    private void addCqlCollectionToRecord(GenericRecord record, CFDefinition.Name name, ByteBuffer buffer) {
+        if (name.type instanceof ListType) {
+            AbstractType elementType = ((ListType) name.type).elements;
+            List<?> list = (List<?>) ListType.getInstance(elementType).compose(buffer);
+            if (needTransformRecord(elementType)) {
+                List<Object> resultList = new ArrayList(list.size());
+                for (Object value: list) {
+                    resultList.add(transformValue(elementType, value));
+                }
+                record.put(name.name.toString(), resultList);
+            } else {
+                record.put(name.name.toString(), list);
+            }
+        } else if (name.type instanceof SetType) {
+            AbstractType elementType = ((SetType) name.type).elements;
+            Set<?> set = (Set<?>)SetType.getInstance(elementType).compose(buffer);
+            if (needTransformRecord(elementType)) {
+                Set<Object> resultSet = new HashSet<Object>(set.size());
+                for (Object value: set) {
+                    resultSet.add(transformValue(elementType, value));
+                }
+                record.put(name.name.toString(), resultSet);
+            } else {
+                record.put(name.name.toString(), set);
+            }
+
+        } else if (name.type instanceof MapType) {
+            AbstractType keyType = ((MapType) name.type).keys;
+            AbstractType valueType = ((MapType) name.type).values;
+            Map<?, ?> map = (Map<?, ?>) MapType.getInstance(keyType, valueType).compose(buffer);
+            if (needTransformRecord(valueType)) {
+                Map<Object, Object> resultMap = new HashMap<Object, Object>(map.size());
+                for (Map.Entry entry: map.entrySet()) {
+                    resultMap.put(entry.getKey(), transformValue(valueType, entry.getValue()));
+                }
+                record.put(name.name.toString(), resultMap);
+            } else {
+                record.put(name.name.toString(), map);
+            }
         }
+    }
 
-        AbstractType<?> type = name.type;
-        Object valueDeserialized = type.compose(value);
+    private boolean needTransformRecord(AbstractType type) {
```

Comment:
`transformValue` also checks specific types. My idea of `needTransformRecord ` is to pre-check and when return false, it can avoid constructing a copy of collection. It may improve performance then I split that function from `transformValue`.  But I agree it is not elegant so far.